Use Istio to block outbound traffic instead of NetworkPolicy

chrisfarms opened the pull request:
## What

* Sets the default outbound traffic policy from Istio sidecars to REGISTRY_ONLY
* Removes the (redundent) default deny egress NetworkPolicy
* Adds an explicit deny egress GlobalNetworkPolicy to prevent communication with the cloudhsm IP unless pod has the `talksToHsm: true` label

## Why

Ultimately; so that pods in managed namespaces can communicate with services running on hosts (like Kiam agent) and to simplify the amount of configuration (like hardcoded apiServer addresses, egress rules, etc.)

## :warning: After merge

Someone will need to:

* delete all the `allow-egress-*` NetworkPolicies from all namespaces in all clusters as our deployment pipeline will not automatically clean these up.
* delete all cloudhsm virtualservices / destinationrules in gsp-system as these are no longer required
* delete all cloudhsm virtualservices from proxy-node chart / namespaces: see: https://github.com/alphagov/verify-proxy-node/pull/325

Commit:
set sidecar outboundTrafficPolicy to REGISTRY_ONLY

we would like the Istio sidecars to block access to external services
that do not have ServiceEntry or VirtualService/Gateway configuration in
place so that we can restrict exfiltration of data without having to use
complex Egress NetworkPolicy rules.

Commit:
remove default deny egress NetworkPolicy

Our "managed namespaces" previously contained NetworkPolicies to deny
egress unless explicitly whitelisted in a NetworkPolicy so that we can
limit data exfiltration.

This causes problems for communication with anything on the "host"
network as NetworkPolicy cannot use podSelector and namespaceSelectors
for pods running with hostNetwork:true

We have since switched to letting Istio manage the egress traffic by
having the sidebar block connections unless explicit ServiceEntries or
VirtualServices/Gateways have been configured. This allows us to focus
on "ingress" based NetworkPolcies making things a bit easier to reason
about and allowing routing to things on the host network via
ServiceEntry

blairboy362 commented:
Here follows what is not so much a well thought-out position as a stream of (somewhat) related thoughts...

My main concern with this change is:

If we add a `ServiceEntry` for, say, the CloudHSM (required for ANY connection to it from a pod we want to connect), how are we preventing arbitrary pods in arbitrary (managed) namespaces from connecting to it (that aren't the ones we want)? As I understand it, following this change, the `ServiceEntry` will add it to the registry and so the envoy side car will allow it through.

We could use the `exportTo` directive on the `ServiceEntry` to only expose it to the current namespace, but that would still mean that every pod in that namespace could connect. And I'm not sure if we can then create multiple `ServiceEntry` resources for the same IP in different namespaces (it probably will work; just want to make sure we test that case too, should this be the route we choose).

Or....I've missed something and this isn't a problem at all.... :man_shrugging:

Or...it might be possible to define `VirtualService` and `DestinationRule` resources along with the `ServiceEntry` to force the `mesh` `Gateway` (the sidecar) to route traffic bound for the HSM IP via the egressgateway (as we are now). We couldn't get that to work across the namespace boundary when we were working on the istio routing previously but I think this is now a slightly different scenario. But that still leaves us in a situation where if those routing rules, for whatever reason, don't get applied correctly, the connection to the HSM is open (albeit still requiring credentials before it's useful).

chrisfarms commented:
The HSM being a cluster-wide/shared resource I think I would lean towards having global (`exportTo: *`) VirtualService / ServiceEntry and removing the proxy-node specific config for this, letting us have a global source label (`talksToHSM` or whatever) to control access... I was not aware that there were problems with this config, so yes this sounds like I should def confirm this first and prep the required changes.

Commit:
deny egress to cloudhsm without talksToHSM label

even though you need credentials to communicate with the cloudhsm that
are only provided to pods that need them, we would still like to
restrict most pods in the cluster from even being able to send traffic.

this adds a GlobalNetworkPolicy that drops traffic to the cloudhsm from any
pods that do not have a talksToHSM=true label.

Commit:
set cloudhsm.enabled=true for hack/lint

when using hack/lint to template out for testing we want it to enable
the cloudhsm stuff too

Commit:
removes redundent cloudhsm virtualservice/gateway

cloudhsm egress limitiation is now controlled by a GlobalNetworkPolicy
and not Istio's egress gateway, so the cloudhsm Gateway and
VirtualService configuration can be removed

chrisfarms commented:
We decided to add a calico `GlobalNetworkPolicy` to further limit communication with the CloudHSM to only pods with a specific label.

Commit:
removes redundent virtualservice/gateway config

the presence of a ServiceEntry is all that is required to whitelist an
external endpoint now that we are restricting egress from the sidecar in
REGISTRY_ONLY mode
